File: apps/ios/Sources/OlloCrypto/IdentityAddress.swift
import Foundation

/// Account address is the long-term identity Ed25519 public key, not a phone.
/// Private key never appears in this encoding.
public enum IdentityAddress {
    public static let prefix = "ollo:user:v1:"
    public static let authProofDomain = "ollo-auth-v1"

    public static func encode(_ ed25519Public: Data) -> String {
        guard ed25519Public.count == 32, ed25519Public.contains(where: { $0 != 0 }) else { return "" }
        return prefix + b64url(ed25519Public)
    }

    public static func parse(_ raw: String) -> Data? {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return nil }
        let payload = s.hasPrefix(prefix) ? String(s.dropFirst(prefix.count)) : s
        guard let bytes = b64urlDecode(payload), bytes.count == 32, bytes.contains(where: { $0 != 0 }) else {
            return nil
        }
        return bytes
    }

    /// Canonical bytes signed to prove possession of the identity Ed25519 key.
    public static func authProof(challengeId: String, nonce: String) -> Data {
        if challengeId.isEmpty || nonce.isEmpty { return Data() }
        var out = Data(authProofDomain.utf8)
        out.append(0)
        out.append(Data(challengeId.utf8))
        out.append(0)
        out.append(Data(nonce.utf8))
        return out
    }

    private static func b64url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .trimmingCharacters(in: CharacterSet(charactersIn: "="))
    }

    private static func b64urlDecode(_ raw: String) -> Data? {
        var t = raw.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        let pad = (4 - t.count % 4) % 4
        if pad > 0 { t += String(repeating: "=", count: pad) }
        return Data(base64Encoded: t)
    }
}

File: apps/android/core/crypto/src/main/java/app/ollo/crypto/IdentityAddress.kt
package app.ollo.crypto

import java.util.Base64

/**
 * Account address is the long-term identity Ed25519 public key, not a phone.
 * Private key never appears in this encoding.
 */
object IdentityAddress {
    const val PREFIX = "ollo:user:v1:"
    const val AUTH_PROOF_DOMAIN = "ollo-auth-v1"

    fun encode(ed25519Public: ByteArray): String {
        if (ed25519Public.size != 32 || ed25519Public.all { it == 0.toByte() }) return ""
        return PREFIX + b64url(ed25519Public)
    }

    fun parse(raw: String): ByteArray? {
        val s = raw.trim()
        if (s.isEmpty()) return null
        val payload = if (s.startsWith(PREFIX)) s.substring(PREFIX.length) else s
        val bytes = b64urlDecode(payload) ?: return null
        if (bytes.size != 32 || bytes.all { it == 0.toByte() }) return null
        return bytes
    }

    /** Canonical bytes signed to prove possession of the identity Ed25519 key. */
    fun authProof(challengeId: String, nonce: String): ByteArray {
        if (challengeId.isEmpty() || nonce.isEmpty()) return ByteArray(0)
        val a = AUTH_PROOF_DOMAIN.toByteArray(Charsets.UTF_8)
        val b = challengeId.toByteArray(Charsets.UTF_8)
        val c = nonce.toByteArray(Charsets.UTF_8)
        return a + byteArrayOf(0) + b + byteArrayOf(0) + c
    }

    private fun b64url(bytes: ByteArray): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

    private fun b64urlDecode(raw: String): ByteArray? {
        return try {
            Base64.getUrlDecoder().decode(raw)
        } catch (_: IllegalArgumentException) {
            null
        }
    }
}
